### ISSUE TITLE
[Microsoft.Build.CentralPackageVersions] Allow per project  package references that are not centrally managed.

### ISSUE THREAD

**Issue by amrmahdi** [microsoft/MSBuildSdks#101]
**SDK: Microsoft.Build.CentralPackageVersions**
**Version: 2.0.36**

The current behavior of Microsoft.Build.CentralPackageVersions forces the consumer to completely opt-in and have all packages version centrally managed. 

It is desirable to have a subset of packages versions be centrally managed, and in the same time have per project packages that are not centrally managed. 

For example, one could have ``Microsoft.Extensions.*`` packages centrally managed in for all projects in a repo, and at the same time continue to reference packages per project as needed.  

A project would look like:

```xml
<PackageReference Include="Microsoft.Extensions.Configuration"  /> <!-- Centrally managed -->
<PackageReference Include="Grpc" Version="1.21.0" /> <!-- regular package reference -->
```

**Comment by yhvicey** [comment]
I have no idea about how to distinguish updated `PackageReference` in `Packages.props` from other normal `PackageReference` without custom tasks, and I guess we have below options:

1. Add custom tasks to distinguish newly updated `PackageReference` in `Pacjages.props`;
    1. This may requires adding C# assembly to this SDK.
1. Allow user define versions in both places (`Packages.props` and project file) and only version in `Packages.props` will take effect; if user wants to override, they must use `VersionOverride` explicity;

Current behavior is:

1. All packages' version must be specified in `Packages.props`;
1. All project files cannot specify any version for any package;
1. If they want to do so, they must use `VersionOverride`.

If we choose option 1, behavior will be look like:

1. All packages' version must be specified at most once, no matter in `Packages.props` or project file;
1. Project files can have their own package versions, until it is also specified in `Packages.props`;
1. If it's already specified in `Packages.props` but need to be overrided, users can use `VersionOverride`.

If we choose option 2, behavior will be look like:

1. Packages can have their version specified in both `Packages.props` and project files;
1. Only version in `Packages.props` will take effect, and there's no error or warning if it's been specified twice;
1. User can override package version by using `VersionOverride`

All above descriptions based on one assumption: This SDK **CAN/WILLING TO** introduce such mechanism. @jeffkl any feedback?

**Comment by jeffkl** [comment]
Sorry I was out on vacation.  There will have to be some convention of marking which PackageReferences are allowed to specify a version.  In your example above:

```xml
<PackageReference Include="Microsoft.Extensions.Configuration"  /> <!-- Centrally managed -->
<PackageReference Include="Grpc" Version="1.21.0" /> <!-- regular package reference -->
```

Either the PackageReference in the project will need to indicate it:

```xml
<PackageReference Include="Microsoft.Extensions.Configuration"  /> <!-- Centrally managed -->
<PackageReference Include="Grpc" Version="1.21.0" CentralVersion="false" /> <!-- regular package reference -->
```
Or the `Packages.props` will need to

```xml
<PackageReference Update="Grpc" AllowVersion="true" /> <!-- regular package reference -->
```

I am not totally convinced this is needed.  The idea is you want all of your package versions in one place.  Getting to specify some versions centrally and others locally seems like it would only cause confusion.  Perhaps I'm missing a use case?

**Comment by ChristophLindemann** [comment]
> I am not totally convinced this is needed. The idea is you want all of your package versions in one place. Getting to specify some versions centrally and others locally seems like it would only cause confusion. Perhaps I'm missing a use case?

I totally agree with the above comment. I my view, you should either manage _all_ versions centrally _OR_ manage them for every project (and use consolidate i VS).

Mixing these approaches, just adds confusion, and you never know where to set the version, and end up with unexpected versions.

If something like this would be implemented, I kindly ask you to make it opt-in, so you actively have to do something to enable such feature.

**Comment by yhvicey** [comment]
I've created a PR (#203) for this to allow users opt-in per-project `PackageReference`. Users can enable this feature by assigning `AllowPackageReferenceWithVersion` to `true` before importing `Microsoft.Build.CentralPackageVersion` sdk, e.g.:

```xml
<Project>
  <PropertyGroup>
    <AllowPackageReferenceWithVersion>true</AllowPackageReferenceWithVersion>
  </PropertyGroup>
  ...
  <Import Project="Sdk.targets" Sdk="Microsoft.Build.CentralPackageVersions" />
</Project>
```

For those who needs this feature, you can add below snippets as a workaround:

In your `Directory.Build.targets` or wherever you think is proper to put this fix:

```xml
<Project>
  <PropertyGroup>
    <AllowPackageReferenceWithVersion>true</AllowPackageReferenceWithVersion>
  </PropertyGroup>

  <!-- #region Workaround -->
  <!-- TODO: Remove this region after https://github.com/microsoft/MSBuildSdks/pull/203 is merged -->
  <ItemGroup Condition=" '$(AllowPackageReferenceWithVersion)' == 'true' ">
    <!-- Store package references with version to avoid check -->
    <_OriginalPackageReferenceWithVersion Include="@(PackageReference->HasMetadata('Version'))" />
    <PackageReference Remove="@(_OriginalPackageReferenceWithVersion)" />
  </ItemGroup>

  <Target Name="RestoreOriginalPackageReferenceWithVersion" AfterTargets="CheckPackageReferences" Condition=" '$(AllowPackageReferenceWithVersion)' == 'true' AND @(_OriginalPackageReferenceWithVersion->Count()) > 0 ">
    <!-- Restore staged package references -->
    <ItemGroup>
      <PackageReference Include="@(_OriginalPackageReferenceWithVersion)" />
    </ItemGroup>
  </Target>
  <!-- #endregion -->

  <!-- Sdk -->
  <Import Project="Sdk.targets" Sdk="Microsoft.Build.CentralPackageVersions" />
</Project>
```

**Comment by teneko** [comment]
Is this still up for a grab, @jeffkl? I use it and find it very useful.

**Comment by jeffkl** [comment]
Microsoft.Build.CentralPackageVersions has been deprecated in favor of NuGet's Central Package Management: https://github.com/microsoft/MSBuildSdks/tree/main/src/CentralPackageVersions#deprecated

If you want to keep using CPV and configure it the way you want, you can just copy all of its logic into your repository's Directory.Build.props and Directory.Build.targets.  But we won't be shipping any more updates to CPV.

**Comment by dfederm** [comment]
Closing since CPV is deprecated
